diffrax/misc/discontinuity.py

```diff
@@ -0,0 +1,62 @@
+from typing import Callable
+
+import diffrax
+import jax
+import jax.numpy as jnp
+from diffrax.custom_types import Array, Scalar
+
+
+def find_root_for_one_delay_one_discontinuity(
+    delay: Callable,
+    discontinuity: float,
+    t0: float,
+    t1: float,
+    y0: Array,
+    y1: Array,
+    atol: float = 1e-6,
+    rtol: float = 1e-3,
+    max_steps: int = 10,
+):
+    cubic_coeffs = diffrax.backward_hermite_coefficients(
+        jnp.array([t0, t1]), jnp.array([y0, y1])
+    )
+    interpol = diffrax.CubicInterpolation(ts=jnp.array([t0, t1]), coeffs=cubic_coeffs)
+    g = lambda t, args: t - delay(t, interpol.evaluate(t), args) - discontinuity
+    root_solver = diffrax.NewtonNonlinearSolver(
+        atol=atol, rtol=rtol, max_steps=max_steps
+    )
+    pred = g(t0, None) * g(t1, None) > 0
+
+    def true_fun(_):
+        return jnp.nan
+
+    def false_fun(t0):
+        # Since steps are pretty small the root finding process
+        # should be almost certainly successful hence only taking the root argument
+        return root_solver(g, t0, None).root
+
+    return jax.lax.cond(pred, true_fun, false_fun, t0)
+
+
+def event_function_root_finding(
+    delays: list[Callable],
+    discontinuities: Array,
+    t0: Scalar,
+    t1: Scalar,
+    y0: Array["state"],  # noqa: F821
+    y1: Array["state"],  # noqa: F821,
+    atol: float = 1e-6,
+    rtol: float = 1e-3,
+    max_steps: int = 10,
+):
+    roots = jnp.full((len(delays), len(discontinuities)), jnp.inf)
+    discontinuity_vmap = jax.vmap(
+        find_root_for_one_delay_one_discontinuity,
+        (None, 0, None, None, None, None, None, None, None),
+    )
+    for i, delay in enumerate(delays):
+        delay_roots = discontinuity_vmap(
+            delay, discontinuities, t0, t1, y0, y1, atol, rtol, max_steps
+        )
+        roots = roots.at[i, :].set(delay_roots)
+    return roots
```

Comment:
FYI this loop may be quite inefficient. Better to do:
```python
roots = []
for delay in delays:
    roots.append(discontinuity_vmap(...))
roots = jnp.stack(roots)
```

Reply:
Will change `discontinuity.py` while i check out the bisection code !